sdk/src/Core/Amazon.Runtime/Pipeline/Handlers/Marshaller.cs

```diff
@@ -106,29 +107,30 @@ private static void SetRecursionDetectionHeader(IDictionary<string, string> head
 
         private static void SetUserAgentHeader(IRequestContext requestContext)
         {
-            var sb = new StringBuilder(requestContext.ClientConfig.UserAgent);
+            var sb = new StringBuilder(256);
+
+            sb.Append(InternalSDKUtils.ReplaceInvalidUserAgentCharacters(requestContext.ClientConfig.UserAgent));
 
             var clientAppId = requestContext.ClientConfig.ClientAppId;
             if (!string.IsNullOrEmpty(clientAppId))
-                sb.AppendFormat(" app/{0}", clientAppId);
+                sb.Append(" app/").Append(InternalSDKUtils.ReplaceInvalidUserAgentCharacters(clientAppId));
 
-            var retryMode = requestContext.ClientConfig.RetryMode.ToString().ToLower();
-            sb.AppendFormat(" cfg/retry-mode#{0}", retryMode);
+            var retryMode = ToUserAgentHeaderString(requestContext.ClientConfig.RetryMode);
+            Debug.Assert(retryMode != requestContext.ClientConfig.RetryMode.ToString().ToLower(), "Invalid RetryMode string.");
```

Comment:
I think doing the whole `requestContext.ClientConfig.RetryMode.ToString().ToLower()` defeats the purpose of having the `ToUserAgentHeaderString` method to remove allocation. The `requestContext.ClientConfig.RetryMode.ToString().ToLower()` is still going to be evaluated in `release` mode. The `Debug.Assert` is always called whether in release or debug mode it is just the evaluated Boolean past into the `Debug.Assert` won't trigger an assert in `release` mode.

Reply:
Have you checked, @normj?

Invocations of methods decorated with `ConditionalAttribute` should not have IL generated if the condition is false.

Check the difference between [this](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUKgBgAJVMA6AEQEsBDAcygHsBnYKgY2YG488A3GmCJUiAXiIEeuPBQCmAIwh0yAQWbNZYYAApEiEQD5xBAJRS8pAJzaqZvEA) and [this](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQcAMACOSB0AIgJYCGA5lAPYDOwxAxjQNzroBupY2x2AvNkys06QgFMARhHL4AgjRpiwwABQBqNbwB8AzAEph6PAE4VxA+iA).

Reply:
@paulomorgado learn something new everyday. Thanks for pointing out my misunderstanding.